src/hooks/useRequest.ts

```diff
@@ -0,0 +1,79 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+
+export interface UseRequestOptions<T> {
+  initialValue?: T;
+  onError?: (err: Error) => void;
+  onSuccess?: (res: T) => void;
+}
+
+export interface UseRequestResponse<T> {
+  loading: boolean;
+  error: Error | undefined;
+  data: T | undefined;
+}
+
+export type UseRequestReturn<T> = [() => Promise<void>, UseRequestResponse<T>];
+
+/**
+ * Provides an API to statefully interact with promises. This hook
+ * prevents state updates if promise has not resolved before component unmounts.
+ * It will return a tuple with a request() function, and an object representing the state
+ */
+export function useRequest<T>(
+  requestFn: () => Promise<T>,
+  options: UseRequestOptions<T> = {}
+): UseRequestReturn<T> {
+  const { initialValue, onSuccess, onError } = options;
+
+  const [loading, setLoading] = useState(false);
+  const [data, setData] = useState<T | undefined>(initialValue);
+  const [error, setError] = useState<Error | undefined>();
+
+  const mountedRef = useRef(false);
+
+  useEffect(() => {
+    mountedRef.current = true;
+
+    return () => {
+      mountedRef.current = false;
+    };
+  }, []);
```

Comment:
I don't understand the point of this function or of `mountedRef` in general. Is this a strategy to delay the execution of the actual request to "next tick" type thing? Feels pretty hacky and why is that necessary?

Reply:
The purpose is so that if a promise hasn't resolved when a component is unmounted, we can prevent `setState` calls on unmounted components, which react will throw warnings for

Reply:
A bit more about this can be found on [this article](https://www.robinwieruch.de/react-warning-cant-call-setstate-on-an-unmounted-component)

Reply:
interesting, ty